File: src/excord-rna.py
import argparse
import pysam

def extract_split_reads(bam_file, output_file, cigar_flags):
    split_reads = []

    with pysam.AlignmentFile(bam_file, "rb") as bam:
        for read in bam:
            # Check if the read has a valid CIGAR string
            if read.cigarstring is not None:
                # Extract split reads based on specified CIGAR flags
                if any(flag in read.cigarstring for flag in cigar_flags):
                    split_reads.append(read)

    # Write split reads to output file
    with open(output_file, "w") as output:
        for read in split_reads:
            output.write(str(read) + "\n")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Extract split reads from a BAM file based on CIGAR flags.")
    parser.add_argument("bam_file", help="Input BAM file")
    parser.add_argument("output_file", help="Output file to store split reads")
    parser.add_argument("-c", "--cigar_flags", nargs="+", default=["N"],
                        help="CIGAR flags to consider for split reads (e.g., -c D I)")
    args = parser.parse_args()

    extract_split_reads(args.bam_file, args.output_file, args.cigar_flags)

File: src/excord-rna.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/brentp/bigly"
	"github.com/brentp/bigly/bamat"
)

func main() {
	// Define command-line flags
	bamPath := flag.String("b", "", "Input BAM file")
	outputFile := flag.String("o", "", "Output file to store split reads")
	cigarFlags := flag.String("c", "", "CIGAR flags to consider for split reads (comma-separated)")
	flag.Parse()

	// Check if required flags are provided
	if *bamPath == "" || *outputFile == "" || *cigarFlags == "" {
		flag.Usage()
		os.Exit(1)
	}

	// Split cigarFlags string into a slice
	cigarFlagSlice := strings.Split(*cigarFlags, ",")

	// Extract split reads based on provided BAM file and CIGAR flags
	err := extractSplitReads(*bamPath, *outputFile, cigarFlagSlice)
	if err != nil {
		log.Fatal(err)
	}
}

func extractSplitReads(bamPath, outputFile string, cigarFlags []string) error {
	// Open BAM file
	bamReader, err := bamat.New(bamPath)
	if err != nil {
		return err
	}
	defer bamReader.Close()

	// Open output file
	outputFileHandle, err := os.Create(outputFile)
	if err != nil {
		return err
	}
	defer outputFileHandle.Close()

	// Write split reads to output file
	splitReads, err := bigly.ExtractSplitReads(bamReader, cigarFlags)
	if err != nil {
		return err
	}

	for _, read := range splitReads {
		fmt.Fprintf(outputFileHandle, "%s\n", read)
	}

	fmt.Println("Split reads extracted successfully and written to", outputFile)
	return nil
}
